File: src/content/3.11/code/scala/snippet12.scala
implicit def freeFunctor[F[_]] = new Functor[FreeF[F, ?]] {
  def fmap[A, B](g: A => B)(fa: FreeF[F, A]): FreeF[F, B] = fa match {
    case FreeF(r) => FreeF {
      new ~>[B => ?, F] {
        def apply[C](bi: B => C): F[C] =
          r(bi compose g)
      }
    }
  }
}

File: src/content/3.11/code/ocaml/snippet12.ml
module FreeFunctorAlt (F : sig
  type 'a f
end) : Functor = struct
  module type F = FreeF_Alt with type 'a f = 'a F.f

  type 'a t = (module F with type a = 'a)

  let fmap (type a' b') f (module FF : F with type a = a') =
    (module struct
      type a = b'
      type 'a f = 'a F.f

      let freeF bi = FF.freeF (fun a -> bi (f a))
    end : F
      with type a = b')
  ;;
end
